ValueError: For an array_value, subvalues must either all be indexed or all excluded from indexes.

Getting `ValueError: For an array_value, subvalues must either all be indexed or all excluded from indexes.` error when querying entities with empty arrays.

Edit: I tried getting around this with projection queries where possible but I need to modify the entities after fetching them, which you can't do with the results of a projection query. So this is a pretty big issue as there is no way around it.

Hi @thisbejim,
Thanks for reporting this issue.

First off, a disclaimer: I am not an expert in Datastore.

That said, I am having a little trouble with a fix for your issue, and it would help if you could provide the code that you think should work. It makes sense to me that an empty array should not raise an exception complaining about mismatched index exclusion of its subelements. That said, I am having trouble making an empty array function at all.

[Here](https://github.com/GoogleCloudPlatform/google-cloud-python/blob/master/datastore/google/cloud/datastore/helpers.py#L136-L142) is our code that checks for consistency on index exclusion:

```python
        if is_list and len(value) > 0:
            exclude_values = set(value_pb.exclude_from_indexes
                                 for value_pb in value_pb.array_value.values)
            if len(exclude_values) != 1:
                raise ValueError('For an array_value, subvalues must either '
                                 'all be indexed or all excluded from '
                                 'indexes.')
```

At first I thought this was a simple fix; simply add `and len(value) > 0:` to the check. However (again, remember, I am unfamiliar with Datastore), when I went to write _tests_ for it, I ran into trouble:

```python
    def test_index_mismatch_ignores_empty_list(self):
        from google.cloud.proto.datastore.v1 import entity_pb2
        from google.cloud.datastore.helpers import _new_value_pb

        _PROJECT = 'PROJECT'
        _KIND = 'KIND'
        _ID = 1234
        entity_pb = entity_pb2.Entity()
        entity_pb.key.partition_id.project_id = _PROJECT
        entity_pb.key.path.add(kind=_KIND, id=_ID)

        array_val_pb = _new_value_pb(entity_pb, 'baz')
        array_pb = array_val_pb.array_value.values

        entity = self._call_fut(entity_pb)
        entity_dict = dict(entity)
        self.assertIsInstance(entity_dict['baz'], list)
```

This raises an exception when I try to run the function under test (which is `entity_from_protobuf`):

```___________________ Test_entity_from_protobuf.test_index_mismatch_ignores_empty_list ___________________

self = <unit_tests.test_helpers.Test_entity_from_protobuf testMethod=test_index_mismatch_ignores_empty_list>

    def test_index_mismatch_ignores_empty_list(self):
        from google.cloud.proto.datastore.v1 import entity_pb2
        from google.cloud.datastore.helpers import _new_value_pb
    
        _PROJECT = 'PROJECT'
        _KIND = 'KIND'
        _ID = 1234
        entity_pb = entity_pb2.Entity()
        entity_pb.key.partition_id.project_id = _PROJECT
        entity_pb.key.path.add(kind=_KIND, id=_ID)
    
        array_val_pb = _new_value_pb(entity_pb, 'baz')
        array_pb = array_val_pb.array_value.values
    
        # unindexed_value_pb1 = array_pb.add()
        # unindexed_value_pb1.integer_value = 10
    
>       entity = self._call_fut(entity_pb)

unit_tests/test_helpers.py:155: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
unit_tests/test_helpers.py:66: in _call_fut
    return entity_from_protobuf(val)
.tox/py35/lib/python3.5/site-packages/google/cloud/datastore/helpers.py:124: in entity_from_protobuf
    value = _get_value_from_value_pb(value_pb)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

value_pb = 

    def _get_value_from_value_pb(value_pb):
        """Given a protobuf for a Value, get the correct value.
    
        The Cloud Datastore Protobuf API returns a Property Protobuf which
        has one value set and the rest blank.  This function retrieves the
        the one value provided.
    
        Some work is done to coerce the return value into a more useful type
        (particularly in the case of a timestamp value, or a key value).
    
        :type value_pb: :class:`.entity_pb2.Value`
        :param value_pb: The Value Protobuf.
    
        :rtype: object
        :returns: The value provided by the Protobuf.
        :raises: :class:`ValueError <exceptions.ValueError>` if no value type
                 has been set.
        """
        value_type = value_pb.WhichOneof('value_type')
    
        if value_type == 'timestamp_value':
            result = _pb_timestamp_to_datetime(value_pb.timestamp_value)
    
        elif value_type == 'key_value':
            result = key_from_protobuf(value_pb.key_value)
    
        elif value_type == 'boolean_value':
            result = value_pb.boolean_value
    
        elif value_type == 'double_value':
            result = value_pb.double_value
    
        elif value_type == 'integer_value':
            result = value_pb.integer_value
    
        elif value_type == 'string_value':
            result = value_pb.string_value
    
        elif value_type == 'blob_value':
            result = value_pb.blob_value
    
        elif value_type == 'entity_value':
            result = entity_from_protobuf(value_pb.entity_value)
    
        elif value_type == 'array_value':
            result = [_get_value_from_value_pb(value)
                      for value in value_pb.array_value.values]
    
        elif value_type == 'geo_point_value':
            result = GeoPoint(value_pb.geo_point_value.latitude,
                              value_pb.geo_point_value.longitude)
    
        elif value_type == 'null_value':
            result = None
    
        else:
>           raise ValueError('Value protobuf did not have any value set')
E           ValueError: Value protobuf did not have any value set
```

This makes me wonder if an empty array is even a valid option. (If I add a value to the array, it works fine.)

Long story short: Can you provide code with what you are trying to do, and your expected result?

--- a/datastore/google/cloud/datastore/helpers.py
+++ b/datastore/google/cloud/datastore/helpers.py
@@ -134,7 +134,7 @@ def entity_from_protobuf(pb):
         # Check if ``value_pb`` was excluded from index. Lists need to be
         # special-cased and we require all ``exclude_from_indexes`` values
         # in a list agree.
-        if is_list:
+        if is_list and len(value) > 0:
             exclude_values = set(value_pb.exclude_from_indexes
                                  for value_pb in value_pb.array_value.values)
             if len(exclude_values) != 1:

--- a/datastore/tests/unit/test_helpers.py
+++ b/datastore/tests/unit/test_helpers.py
@@ -191,6 +191,28 @@ class Test_entity_from_protobuf(unittest.TestCase):
         self.assertEqual(len(inside_entity), 1)
         self.assertEqual(inside_entity[INSIDE_NAME], INSIDE_VALUE)
 
+    def test_index_mismatch_ignores_empty_list(self):
+        from google.cloud.datastore_v1.proto import entity_pb2
+
+        _PROJECT = 'PROJECT'
+        _KIND = 'KIND'
+        _ID = 1234
+
+        array_val_pb = entity_pb2.Value(
+            array_value=entity_pb2.ArrayValue(values=[]))
+
+        entity_pb = entity_pb2.Entity(
+            properties={
+                'baz': array_val_pb,
+            },
+        )
+        entity_pb.key.partition_id.project_id = _PROJECT
+        entity_pb.key.path.add(kind=_KIND, id=_ID)
+
+        entity = self._call_fut(entity_pb)
+        entity_dict = dict(entity)
+        self.assertEqual(entity_dict['baz'], [])
+
 
 class Test_entity_to_protobuf(unittest.TestCase):
 

